ganga/GangaCore/Lib/Localhost/Localhost.py

```diff
@@ -231,7 +236,11 @@ def preparejob(self, jobconfig, master_input_sandbox):
         script = script.replace('###ENVIRONMENT###', repr(environment))
         script = script.replace('###WORKDIR###', repr(workdir))
         script = script.replace('###INPUT_DIR###', repr(job.getStringInputDir()))
-
+        script = script.replace('###VIRTUALIZATION###', repr(virtualization.__class__.__name__))
+        script = script.replace('###VIRTUALIZATIONIMAGE###', repr(virtualization.imageUrl))
+        script = script.replace('###VIRTUALIZATIONMODE###', repr(virtualization.mode) if hasattr(virtualization, 'mode') else repr(None)  )
```

Comment:
If I understand the code correctly then the default value for the `Job.virtualization` attribute is None. If a user is not using virtualisation and just submitting a normal job then the `Job.virtualization` attribute will have this default value. The above lines will then break with an `AttributeError` right as `None` clearly doesn't have these attributes?

Reply:
Fixed

Reply:
I don't think I was perhaps clear. The getters in the `IVirtualization` class haven't changed the fact that `LocalHost.py` either has to have knowledge of the virtualisation attributes type or have a certain level of if/elsery to catch all cases. What I was suggesting would see these lines here reduced to the following:

```python
if job.virtualization is not None:
    script = job.virtualization.modify_script(script)
```
Where you can call the actual method what you like.

---
To achieve this I was suggesting the base class will have an abstract method like so:
```python
class IVirtualization:
...
def modify_script(self, script):
        return script.replace('###VIRTUALIZATION###', repr(self.__class__.__name__))\
                     .replace('###VIRTUALIZATIONIMAGE###', repr(self.imageUrl))
...
```

And the derived classes like `Docker` for example would override this hence using polymorphism instead of if/else/else/else ... for however many derived classes we end up making:

```python
Class Docker(IVirtualization):
...

def modify_script(self, script):
    return super(Docker, self).modify_script(script)\
                              .replace('###VIRTUALIZATIONMODE###', repr(self.mode))
...
```

This way there is no implicit knowledge of the type of the virtualization within `LocalHost.py`, no if /elsery  and everything is self contained, modular and scalable.

Reply:
done